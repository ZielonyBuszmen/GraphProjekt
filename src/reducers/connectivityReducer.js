/**
 * 31. Testowanie silnej spójności grafu
 *
 * Reducer zawiera wszystkie funkcje obliczające silną spójność grafu.
 * Zmienia w modelu flagę `isGraphConnected`, gdzie `true` oznacza graf silnie spójny, `false` mówi o braku takiej spójności,
 * a `null` oznacza stan początkowy flagi
 */
import {createEmptyArray, Stack} from '../helpers';

export const CHECK_GRAPH_CONNECTIVITY = 'connectivity/CHECK_GRAPH_CONNECTIVITY';
export const RESET_STATUS = 'RESET_STATUS';
export const BLOCK_GRAPH_INPUTS = 'BLOCK_GRAPH_INPUTS';

/**
 * Główna funkcja, którą dostarcza ten reducer to `connectivity()`.
 *
 * - Gdy zostanie wywołana akcja CHECK_GRAPH_CONNECTIVITY i gdy graf jest silnie spójny,
 *   funkcja zmienia flagę `isGraphConnected` na `true`. W przeciwnym razie na `false`.
 * - Gdy zostanie wywołana akcja RESET_STATUS, flaga jest ustawiona na `null` (wartość domyślna tej flagi)
 *
 * Gdy akcja to CHECK_GRAPH_CONNECTIVITY:
 *   Jako pierwszy parametr przyjmuje aktualny stan aplikacji (a dokładniej wycinek stanu odpowiadjącego za spójność grafu)
 *   Jako drugi parametr dostaje akcję, wraz z grafem, którego spójność ma sprawdzać.
 * Przyjmowana akcja ma taką postać:
 * action = {
 *   type: CHECK_GRAPH_CONNECTIVITY,
 *   graph: macierz_sasiedztwa
 * }
 * gdzie `macierz_sasiedztwa` może wyglądać w taki sposób:
 * [
 *   [1, 0, 1],
 *   [0, 1, 0],
 *   [1, 1, 1],
 * ];
 *
 * Gdy akcja to RESET_STATUS:
 * Reducer może przyjąć też drugą akcję, która resetuje informację o sprawdzanym grafie (czyli ustawia flagę na `null`):
 *  action = {
 *   type: RESET_STATUS,
 * }
 */
export default function connectivity(state = [], action) {
  switch (action.type) {
    case CHECK_GRAPH_CONNECTIVITY:
      const newState = {...state};
      newState.isGraphConnected = checkStrongConnectivityOfGraph(action.graph);
      return newState;

    case RESET_STATUS:
      const resettedState = {...state};
      resettedState.isGraphConnected = null;
      return resettedState;

    case BLOCK_GRAPH_INPUTS:
      const blockedState = {...state};
      blockedState.blockGraphInputs = action.block;
      return blockedState;

    default:
      return state;
  }
}

/**
 * Sprawdza silną spójność grafu.
 * Funkcja przyjmuje graf w formie macierzy sąsiedztwa.
 *
 * @param graph
 * @returns {boolean} - true, jeśli graf jest silnie spójny
 */
function checkStrongConnectivityOfGraph(graph) {
  for (const node in graph) {
    const result = checkConnectivityOfGraph(graph, node);
    if (result === false) {
      return false;
    }
  }
  return true;
}

/**
 * Sprawdza spójność grafu
 * Funkcja przyjmuje graf w formie macierzy sąsiedztwa.
 * Zwraca `true`, jeśli graf jest spójny, bądź false w przeciwnym wypadku
 *
 * @param graph - graf w formie macierzy sąsiedztwa
 * @param startedNode - wierchołek, z którego startujemy, domyślnie pierwszy (czyli 0)
 * @return boolean
 */
function checkConnectivityOfGraph(graph, startedNode = 0) {
  const visited = createEmptyArray(graph.length, false);
  const stack = new Stack();
  let visitedCounter = 0;

  stack.push(startedNode);
  visited[startedNode] = true;

  while (!stack.empty()) {
    const v = stack.pop();
    visitedCounter++;

    for (const neighbour in graph[v]) {
      if (graph[v][neighbour] === 0) {
        continue;
      }
      if (visited[neighbour]) {
        continue;
      }
      stack.push(neighbour);
      visited[neighbour] = true;
    }
  }
  return visitedCounter === graph.length;
}
