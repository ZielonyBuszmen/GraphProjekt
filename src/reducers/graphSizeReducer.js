export const CHANGE_GRAPH_SIZE = 'CHANGE_GRAPH_SIZE';

export default function graphSizeReducer(state = [], action) {
  switch (action.type) {
    case CHANGE_GRAPH_SIZE:
      return action.size;
    default:
      return state;
  }
}
