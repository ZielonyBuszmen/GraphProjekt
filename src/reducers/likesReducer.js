// tutaj wszystkie typy akcji
export const INCREMENT_LIKES = 'posts/INCREMENT_LIKES';

// przyk≈Çadowy reducer
function likesReducer(state = [], action) {
  switch (action.type) {
    case INCREMENT_LIKES:
      const newState = {...state};
      if (action.ktory === 'jeden') {
        newState.jeden++;
      } else if (action.ktory === 'dwa') {
        newState.dwa++;
      }
      return newState;
    default:
      return state;
  }

}

export default likesReducer;