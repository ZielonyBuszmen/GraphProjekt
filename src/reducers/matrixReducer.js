
export const MACIERZ_SASIEDZTWA_CHANGED = 'matrix/MACIERZ_SASIEDZTWA_CHANGED';
export const LISTY_SASIEDZTWA_CHANGED = 'matrix/LISTY_SASIEDZTWA_CHANGED';


function matrixReduxer(state = [], action) {
  switch (action.type) {
    case MACIERZ_SASIEDZTWA_CHANGED: // graf w formie macierzy sąsiedztwa się zmienił - wystarczy go zapisać do store
      const newMatrix = action.graph;
      return newMatrix;
    case LISTY_SASIEDZTWA_CHANGED: // graf w formie listy sąsiedztw się zmienił - trzeba go zamienić na macierz sąsiedztwa i zapisać w store
      // todo - zamiana grafu w formie listy sąsiedztw na macierz sąsiedztwa
    default:
      return state;
  }
}

export default matrixReduxer;