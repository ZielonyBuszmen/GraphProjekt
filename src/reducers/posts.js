// tutaj wszystkie typy akcji
export const INCREMENT_LIKES = 'posts/INCREMENT_LIKES';

// przyk≈Çadowy reducer
function posts(state = [], action) {
  switch (action.type) {
    case INCREMENT_LIKES:
      const newState = [...state];
      newState[action.index].likes++;
      return newState;
    default:
      return state;
  }

}

export default posts;