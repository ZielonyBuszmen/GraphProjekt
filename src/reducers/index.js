import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'

import likesReducer from './likesReducer';
import connectivityReducer from './connectivityReducer';
import matrixReduxer from './matrixReducer';
import graphSizeReducer from './graphSizeRecuder';

const rootReducer = combineReducers({
  routing: routerReducer,
  likes: likesReducer,
  connectivity: connectivityReducer,
  matrix: matrixReduxer, // tutaj będzie przechowywany graf w formie macierzy sąsiedztwa
  graphSize: graphSizeReducer,
});

export default rootReducer;