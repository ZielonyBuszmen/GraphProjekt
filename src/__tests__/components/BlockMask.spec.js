import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import '../../enzymeConfiguration';

import {
  BlockMask,
  mapDispatchToProps,
  mapStateToProps,
} from '../../components/BlockMask';
import { BLOCK_GRAPH_INPUTS } from '../../reducers/connectivityReducer';

describe('BlockMask', () => {
  it('should render without props', function () {
    const component = shallow(<BlockMask/>);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render blocked mask', function () {
    const component = shallow(<BlockMask blockGraphInputs/>);
    expect(component.find('.text-center')).toHaveLength(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should unlock mask', function () {
    const unlockGraphInputsMock = jest.fn(() => {});
    const component = shallow(<BlockMask blockGraphInputs unlockGraphInputs={unlockGraphInputsMock}/>);
    component.find('Button').simulate('click');
    expect(unlockGraphInputsMock.mock.calls.length).toBe(1);
  });
});

describe('mapStateToProps', function () {
  it('should return object with state', function () {
    const state = {
      connectivity: {
        blockGraphInputs: true
      }
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({blockGraphInputs: true});
  });
});

describe('mapDispatchToProps', function () {
  it('should contain dispatches', function () {
    const dispatch = (element) => {
      expect(element).toEqual({
        type: BLOCK_GRAPH_INPUTS,
        block: false,
      })
    };
    const result = mapDispatchToProps(dispatch);
    expect(result.hasOwnProperty('unlockGraphInputs')).toBeTruthy();
    result.unlockGraphInputs();
  });
});


