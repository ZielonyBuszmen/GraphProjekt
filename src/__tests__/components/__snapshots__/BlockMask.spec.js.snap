// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockMask should render blocked mask 1`] = `
<div
  className="block"
>
  <Row
    tag="div"
  >
    <Col
      className="text-center"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Button
        color="secondary"
        onClick={[Function]}
        size="lg"
        style={
          Object {
            "backgroundColor": "#ff2953",
            "color": "#ffffff",
          }
        }
        tag="button"
      >
        Edycja
      </Button>
    </Col>
  </Row>
</div>
`;

exports[`BlockMask should render without props 1`] = `""`;
